clients/lsp-netbeans.el

```diff
@@ -0,0 +1,129 @@
+;; - Launch JVM debug config
+
+;;; Code:
+
+(require 'lsp-mode)
+(require 'dap-mode)
+
+(defcustom lsp-netbeans-download-url "https://marketplace.visualstudio.com/_apis/public/gallery/publishers/ASF/vsextensions/apache-netbeans-java/12.4.301/vspackage"
+  "URL to download the Apache Netbeans LSP server from"
+  :group 'lsp-netbeans
+  :type 'string)
+
+(defcustom lsp-netbeans-install-dir (f-join lsp-server-install-dir "asf.apache-netbeans-java")
+  "Apache Netbeans language server installation dir"
+  :group 'lsp-netbeans
+  :type 'string)
+
+(defun lsp-netbeans-server-command (main-port)
+  `(,(f-join lsp-netbeans-install-dir "run.sh")
+    "--start-java-debug-adapter-server=listen:0"
+    ,(format "--start-java-language-server=listen:%d" main-port)))
+
+(defun lsp-netbeans--install-server (client callback error-callback update?)
+  (unless (and (not update?)
+               (f-exists? lsp-netbeans-install-dir))
+    (make-directory lsp-netbeans-install-dir t)
+    (let ((download-path (f-join lsp-netbeans-install-dir "vspackage")))
+      (lsp-download-install
+       (lambda ()
+         (let ((run-script-path (f-join lsp-netbeans-install-dir "run.sh")))
+           (with-temp-file run-script-path
+             (insert "#!/bin/bash
+                      cd `dirname $0`
+                      cd extension
+                      node out/nbcode.js $@ >&2")
+             (executable-make-buffer-file-executable-if-script-p)))
+         (message "Done downloading Netbeans LSP server"))
+       (lambda (err) (error err))
+       :url lsp-netbeans-download-url
+       :store-path download-path
+       :decompress :zip))))
+
+(lsp-register-client
+ (make-lsp-client
+  :new-connection (lsp-tcp-connection 'lsp-netbeans-server-command)
+  :activation-fn (lsp-activate-on "java")
+  :server-id 'netbeans
+  :initialization-options (list :nbcodeCapabilities
+                                (list
+                                 :testResultsSupport (lsp-json-bool nil)
+                                 :statusBarMessageSupport (lsp-json-bool nil)
+                                 :wantsGroovySupport (lsp-json-bool nil)))
+  :priority 10
+  :multi-root t
+  :download-server-fn #'lsp-netbeans--install-server))
+
+(defun dap-netbeans--populate-attach-args (conf)
+  (dap--put-if-absent conf :hostName (read-string "Enter host: " "localhost"))
+  (dap--put-if-absent conf :port (read-string "Enter port: " "8000"))
+  (dap--put-if-absent conf :host "localhost")
+  (dap--put-if-absent conf :name (format "%s(%s)" (plist-get conf :host) (plist-get conf :port)))
+  conf)
+
+(defun dap-netbeans--populate-default-args (conf)
+  (setq conf (plist-put conf :type "java8+"))
+
+  (setq conf (pcase (plist-get conf :request)
+               ("attach" (dap-netbeans--populate-attach-args conf))
+               (_ (dap-netbeans--populate-attach-args conf))))
+
+  (plist-put conf :debugServer (with-current-buffer (get-buffer "*netbeans::stderr*")
```

Comment:
Relying on the `*netbeans::stderr*` buffer name here also seems very fragile to me.